File: week-5/lecture/5-quick-sort/quick-sort.cpp
/**
 * @file quick-sort.cpp
 * @brief C++ version of the week 5 quick-sort lecture example.
 */

#include <iostream>
#include <vector>

using namespace std;

/**
 * @brief Place the pivot in its final quick-sort position.
 * @param array vector<int>& List of integers being sorted in place.
 * @param low int Left boundary of the current partition.
 * @param high int Right boundary of the current partition. The pivot is array[high].
 * @return int Final pivot index after partitioning.
 */
int partition(vector<int>& array, int low, int high) {
    int pivot = array[high];
    int smaller = low - 1;

    for (int index = low; index < high; ++index) {
        if (array[index] <= pivot) {
            ++smaller;
            swap(array[smaller], array[index]);
        }
    }

    swap(array[smaller + 1], array[high]);
    return smaller + 1;
}

/**
 * @brief Sort the list in ascending order using quick sort.
 * @param array vector<int>& List of integers to sort in place.
 * @param low int Left boundary of the current recursive range.
 * @param high int Right boundary of the current recursive range.
 * @return void This function updates array in place and does not return a value.
 */
void quickSort(vector<int>& array, int low, int high) {
    if (low < high) {
        int pivot_index = partition(array, low, high);
        quickSort(array, low, pivot_index - 1);
        quickSort(array, pivot_index + 1, high);
    }
}

/**
 * @brief Run the quick-sort lecture example.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    vector<int> data = {8, 7, 2, 1, 0, 9, 6};
    cout << "Unsorted Array\n";
    for (int value : data) {
        cout << value << ' ';
    }
    cout << '\n';

    quickSort(data, 0, static_cast<int>(data.size()) - 1);

    cout << "Sorted Array in Ascending Order:\n";
    for (int value : data) {
        cout << value << ' ';
    }
    cout << '\n';
    return 0;
}

File: week-5/lecture/5-quick-sort/quick-sort.py
"""Quick Sort example.

This file belongs to the week 5 quick sort lecture example. It uses partition, quickSort to illustrate the main idea with a small runnable example.
"""

# Quick sort in Python

# function to find the partition position
def partition(array, low, high):

    # choose the rightmost element as pivot
    """Place the pivot in its final quick-sort position and return its index.

    Args:
        array (list): The list being processed by the algorithm.
        low (int): The low input used by this function.
        high (int): The high input used by this function.

    Returns:
        int: The integer result produced by the algorithm.
    """
    pivot = array[high]

    # pointer for greater element
    i = low - 1

    # traverse through all elements
    # compare each element with pivot
    for j in range(low, high):
        if array[j] <= pivot:
            # if element smaller than pivot is found
            # swap it with the greater element pointed by i
            i = i + 1

            # swapping element at i with element at j
            (array[i], array[j]) = (array[j], array[i])

    # swap the pivot element with the greater element specified by i
    (array[i + 1], array[high]) = (array[high], array[i + 1])

    # return the position from where partition is done
    return i + 1

# function to perform quicksort
def quickSort(array, low, high):
    """Sort the list in ascending order using quick sort.

    Args:
        array (list): The list being processed by the algorithm.
        low (int): The low input used by this function.
        high (int): The high input used by this function.

    Returns:
        None: This function updates data or prints results and does not return a value.
    """
    if low < high:

        # find pivot element such that
        # element smaller than pivot are on the left
        # element greater than pivot are on the right
        pi = partition(array, low, high)

        # recursive call on the left of pivot
        quickSort(array, low, pi - 1)

        # recursive call on the right of pivot
        quickSort(array, pi + 1, high)


data = [8, 7, 2, 1, 0, 9, 6]
print("Unsorted Array")
print(data)

size = len(data)

quickSort(data, 0, size - 1)

print('Sorted Array in Ascending Order:')
print(data)